plugins/core/hu.bme.mit.gamma.expression.model/src/hu/bme/mit/gamma/expression/util/ExpressionModelValidator.java

```diff
@@ -650,4 +671,176 @@ else if (variableDeclarationType instanceof TypeReference &&
+			for (FieldDeclaration fieldDeclaration : recordTypeDefinition.getFieldDeclarations()) {
+				// TypeReference
+				Type fieldType = (Type) fieldDeclaration.getType();
+				if (fieldType instanceof TypeReference) {
+					TypeReference fieldTypeReference = (TypeReference) fieldType;
+					TypeDeclaration fieldReferencedTypeDeclaration = fieldTypeReference.getReference();
+					// equal with checked record
+					if (fieldReferencedTypeDeclaration == typeDeclaration) {
+						validationResultMessages.add(new ValidationResultMessage(ValidationResult.ERROR,
+								"Record cannot store itself either directly or indirectly! " +
+								visitedNodes.get(0).getName().toUpperCase() + " stores " +
+								typeDeclaration.getName().toUpperCase(),
+								new ReferenceInfo(ExpressionModelPackage.Literals.DECLARATION__TYPE, null)));
+					}
+					// check - if it doesn't equal with checked record and if it isn't visited record
+					else if (!visitedNodes.contains(fieldReferencedTypeDeclaration)) {
+						visitedNodes.add(0, fieldReferencedTypeDeclaration);
+						validationResultMessages.addAll(checkRecordSelfReferenceHelp(typeDeclaration, visitedNodes));
+					}
+				}
+			}
+		}
+		return validationResultMessages;
+	}
+
+	public Collection<ValidationResultMessage> checkRationalLiteralExpression(RationalLiteralExpression expression) {
+		Collection<ValidationResultMessage> validationResultMessages = new ArrayList<ValidationResultMessage>();
+		try {
+			// check denominator
+			if (expression.getDenominator().intValue() == 0) {
+				validationResultMessages.add(new ValidationResultMessage(ValidationResult.ERROR,
+						"The denominator cannot be zero.",
+						new ReferenceInfo(ExpressionModelPackage.Literals.RATIONAL_LITERAL_EXPRESSION__DENOMINATOR, null)));
+			}
+		} catch (Exception exception) {
+			// There is a type error on a lower level, no need to display the error message on this level too
+		}
+		return validationResultMessages;
+	}
+
+	public Collection<ValidationResultMessage> checkRecordLiteralExpression(RecordLiteralExpression expression) {
+		Collection<ValidationResultMessage> validationResultMessages = new ArrayList<ValidationResultMessage>();
+		
+		// find RecordTypeDefinition
+		TypeDeclaration typeDeclaration = expression.getTypeDeclaration();
+		Type type = typeDeclaration.getType();
+		RecordTypeDefinition recordTypeDefinition = (RecordTypeDefinition) type;
+		
+		// check all FieldDeclaration and all FieldAssignment
+		for (FieldDeclaration rTypeField : recordTypeDefinition.getFieldDeclarations()) {
+			int counter = 0;
+			for (FieldAssignment rLiFieldAssignment : expression.getFieldAssignments()) {
+				FieldReferenceExpression fieldReferenceExpression = rLiFieldAssignment.getReference();
+				FieldDeclaration fieldDeclaration = fieldReferenceExpression.getFieldDeclaration();
+				// same field names
+				if (fieldDeclaration.getName() == rTypeField.getName()) {
```

Comment:
I think instead of `fieldDeclaration.getName() == rTypeField.getName()`, `fieldDeclaration == rTypeField`  could be used.